### PR TITLE
On Device Sampling

### PR DESCRIPTION
**Pull request by quic-sanising**
### 📌 Overview

This PR introduces **On Device Sampling** for `QEffForCausalLM` models, enabling sampling operations to be executed directly on the **QAIC device** rather than the host CPU. This enhancement significantly reduces host-device communication overhead and improves inference throughput and scalability.

</br>
</br>

### 🚀 Motivation

Traditionally, sampling (e.g., greedy, top-k, top-p) is performed on the host CPU after logits are returned from the device. This approach incurs:

- High PCIe traffic due to large logits tensors `[batch_size, vocab_size]`
- Latency bottlenecks from CPU-bound sampling logic
- Limited scalability due to CPU thread constraints

**On Device Sampling** addresses these issues by:

- Performing sampling directly on the QAIC device
- Returning only the selected next tokens `[batch_size, 1]`
- Leveraging the device’s parallelism and optimized compute paths

</br>
</br>

### ⚙️ Supported Sampling Strategies

The following sampling techniques are now supported natively on the QAIC device:

1. **Repetition Penalty**​: Penalize tokens that have appeared in the prompt and the generated text so far. Values > 1 encourage the model to use new tokens, while values < 1 encourage the model to repeat tokens.​

2. **Presence Penalty**​: Penalize tokens that are present in the generated text so far. Values > 0 encourage the model to use new tokens, while values < 0 encourage the model to repeat tokens.

3. **Temperature Scaling​**: Adjust the sharpness of the logits distribution. Lower values make the model more deterministic, while higher values make the model more random. Zero means greedy sampling.​

4. **Top K​**: Sample from the `k` largest tokens by value.

5. **Top P​**: Sample from the smallest set of tokens whose cumulative probability is greater than or equal to `p`.​ Must be in (0, 1]. Set to 1 to consider all tokens.​

6. **Min P​**: Represents the minimum probability for a token to be considered, relative to the probability of the most likely token. Must be in [0, 1]. Set to 0 to disable this.​

7. **Greedy Sampling​**: Choose the token with highest value.

8. **Random Sampling**: Choose a token randomly with its probability of being chosen given by its value.

</br>
</br>

### 🛠️ Implementation Details

- **Sampler Integration**: Sampling logic is injected via `include_sampler=True` during model loading. No changes to the model architecture are required.

- **Memory Optimization**: Two scratch buffers of shape `[batch_size, vocab_size]` are used to track token occurrences for applying repetition and presence penalties efficiently on-device.

- **Performance Gains**:
  - Reduced PCIe traffic (logits → tokens)
  - Higher throughput via device-level parallelism
  - Scalable to 64+ concurrent inference streams

</br>
</br>

### 🧪 Usage

```python
from QEfficient import QEFFAutoModelForCausalLM as AutoModelForCausalLM

# Load model with On Device Sampler enabled
qeff_model = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-3.1-8B",
    include_sampler=True,
    return_pdfs=False,
)

# Compile as usual
qeff_model.compile(
    prefill_seq_length=128,
    ctx_len=256,
    full_batch_size=16,
    num_devices=4,
    num_speculative_tokens=0,
    mxint8_kv_cache=True,
    mxfp6_matmul=True,
)
```

To disable On Device Sampling and revert to host-side sampling, simply set `include_sampler=False`.

### CODE REVIEW

File: QEfficient/transformers/models/modeling_auto.py [hunk 1]
```diff
@@ -75,7 +76,7 @@ def __repr__(self) -> str:
 
     @classmethod
     @with_replaced_quantizers
-    def from_pretrained(cls, pretrained_model_name_or_path: str, is_tlm: bool = False, *args, **kwargs):
+    def from_pretrained(cls, pretrained_model_name_or_path: str, is_tlm: bool = False, include_sampler: bool = False, return_pdfs: bool = False, *args, **kwargs):
```
Comment (quic-hemagnih) [inline QEfficient/transformers/models/modeling_auto.py:79]: Should we make them optional parameters

File: QEfficient/transformers/models/modeling_auto.py [hunk 2]
```diff
@@ -1317,8 +1322,14 @@ def __init__(
         if is_tlm:
             # TODO: It is possible to always apply this transform and make value of indices as last indices by default in PyTorch
             self.model, transformed = SpDTransform.apply(self.model)
+            self.model.return_pdfs = True
```
Comment (quic-hemagnih) [inline QEfficient/transformers/models/modeling_auto.py:1325]: where is the code for handling is_tlm == FALSE?

File: QEfficient/transformers/models/modeling_auto.py [hunk 3]
```diff
@@ -1454,6 +1475,48 @@ def export(self, export_dir: Optional[str] = None) -> str:
...
+            dynamic_axes["presence_penalties"] = {0: "batch_size"}
+
+            example_inputs["temperatures"] = torch.ones((bs, 1), dtype=torch.float)
+            dynamic_axes["temperatures"] = {0: "batch_size"}
+
+            example_inputs["top_ks"] = torch.randint(1, max_top_k_ids, size=(bs, 1)).to(torch.int32)
+            dynamic_axes["top_ks"] = {0: "batch_size"}
+
+            example_inputs["top_ps"] = torch.ones((bs, 1), dtype=torch.float) * 0.80
+            dynamic_axes["top_ps"] = {0: "batch_size"}
```
Comment (quic-hemagnih) [inline QEfficient/transformers/models/modeling_auto.py:1512]: Can we define constants for 0.80 and 0.99

File: QEfficient/transformers/models/pytorch_transforms.py [hunk 4]
```diff
@@ -410,6 +410,36 @@ def apply(cls, model: nn.Module) -> Tuple[nn.Module, bool]:
...
+
+    # supported architectures
+    _module_mapping = {
+        # Llama
+        QEffLlamaForCausalLM,
+    }
+
+    @classmethod
+    def apply(cls, model: nn.Module) -> Tuple[nn.Module, bool]:
+        transformed = False
```
Comment (quic-hemagnih) [inline QEfficient/transformers/models/pytorch_transforms.py:431]: Please add doc string

File: QEfficient/transformers/models/modeling_auto.py (hunk 2 above)
Comment (quic-sanising) [reply QEfficient/transformers/models/modeling_auto.py]: This is removed.

File: QEfficient/transformers/models/modeling_auto.py (hunk 3 above)
Comment (quic-sanising) [reply QEfficient/transformers/models/modeling_auto.py:1555]: Done

File: QEfficient/transformers/models/modeling_auto.py [hunk 5]
```diff
@@ -1423,6 +1444,7 @@ def model_hash(self) -> str:
         mhash.update(to_hashable(self.model.config.to_diff_dict()))
         mhash.update(to_hashable({"continuous_batching": self.continuous_batching}))
         mhash.update(to_hashable({"is_tlm": self.is_tlm}))
```
[the remark this answers is not in the document]
Comment (quic-sanising) [reply QEfficient/transformers/models/modeling_auto.py:1453]: Done

File: QEfficient/transformers/models/modeling_auto.py [hunk 6]
```diff
@@ -1610,12 +1694,23 @@ def compile(
         if prefill_only is None or prefill_only or prefill_seq_len == 1:
             specializations.append(
                 self.build_prefill_specialization(
-                    prefill_seq_len, ctx_len, batch_size, kv_cache_batch_size, full_batch_size
+                    prefill_seq_len=prefill_seq_len,
+                    ctx_len=ctx_len,
+                    batch_size=batch_size,
+                    kv_cache_batch_size=kv_cache_batch_size,
+                    full_batch_size=full_batch_size,
+                    max_top_k_ids=constants.Constants.MAX_TOP_K_IDS if self.include_sampler else None,
```
[the remark this answers is not in the document]
Comment (quic-sanising) [reply QEfficient/transformers/models/modeling_auto.py]: Done

File: QEfficient/transformers/models/modeling_auto.py [hunk 7]
```diff
@@ -1331,10 +1341,20 @@ def __init__(
...
         self.is_tlm = transformed
         self.pretrained_model_name_or_path = kwargs.get("pretrained_model_name_or_path", None)
 
+        # ---Sampling---
+        # Note: SamplerTransform should be applied after all other transforms
+        # are done. The role of the sampler is to just add nodes at the output of the
+        # previous transform function.
+        self.model, transformed = SamplerTransform.apply(self.model, qaic_config, **kwargs)
+        if self.is_tlm:
+            self.model.qaic_config["return_pdfs"] = True
```
Comment (ochougul) [inline QEfficient/transformers/models/modeling_auto.py:1356]: can this also be done inside the transform

File: QEfficient/transformers/models/modeling_auto.py [hunk 8]
```diff
@@ -1484,6 +1519,54 @@ def export(self, export_dir: Optional[str] = None) -> str:
...
+
+            max_top_k_ids = self.model.qaic_config.get("max_top_k_ids", constants.ONNX_EXPORT_EXAMPLE_MAX_TOP_K_IDS)
+            example_inputs["top_ks"] = torch.randint(1, max_top_k_ids, size=(bs, 1)).to(torch.int32)
+            dynamic_axes["top_ks"] = {0: "batch_size"}
+
+            example_inputs["top_ps"] = torch.ones((bs, 1), dtype=torch.float) * constants.ONNX_EXPORT_EXAMPLE_TOP_PS
+            dynamic_axes["top_ps"] = {0: "batch_size"}
+
+            example_inputs["min_ps"] = torch.ones((bs, 1), dtype=torch.float) * constants.ONNX_EXPORT_EXAMPLE_MIN_PS
+            dynamic_axes["min_ps"] = {0: "batch_size"}
```
Comment (ochougul) [inline QEfficient/transformers/models/modeling_auto.py:1565]: better to move this code to another func, just for readability

**Review by ochougul — CHANGES_REQUESTED** [review]

File: QEfficient/transformers/models/modeling_auto.py (hunk 8 above)
Comment (quic-sanising) [reply QEfficient/transformers/models/modeling_auto.py]: Done

File: QEfficient/transformers/models/modeling_auto.py (hunk 7 above)
Comment (quic-sanising) [reply QEfficient/transformers/models/modeling_auto.py:1356]: No, `self.is_tlm` is not accessible inside `SamplerTransform`

File: QEfficient/transformers/models/modeling_auto.py [hunk 9]
```diff
@@ -1467,7 +1496,13 @@ def export(self, export_dir: Optional[str] = None) -> str:
...
                 2: "ctx_len",
             }
-        output_names = ["logits"]
+        output_names = []
+        if self.model.qaic_config is not None and self.model.qaic_config.get("include_sampler", False):
+            if self.model.qaic_config.get("return_pdfs", False):
+                output_names.append("probs")
+            output_names.append("next_tokens")
+        else:
+            output_names.append("logits")
```
[the remark this answers is not in the document]
Comment (quic-sanising) [reply QEfficient/transformers/models/modeling_auto.py:1505]: Do you want to go ahead with that approach?

[the remark this answers is not in the document]
Comment (quic-sanising) [reply QEfficient/transformers/models/modeling_auto.py:1505]:
```py
    def get_sampling_inputs_and_outputs(
        self,
        example_inputs: Dict[str, torch.Tensor],
        output_names: List[str],
        dynamic_axes: Dict[str, Dict[int, str]],
    ):
        if self.model.qaic_config.get("return_pdfs", False):
            output_names.append("probs")
        output_names.append("next_tokens")
        output_names.remove("logits") 
    
    ...
```

[the remark this answers is not in the document]
Comment (quic-sanising) [reply QEfficient/transformers/models/modeling_auto.py:1505]: I tried this change, but `remove()` is causing problems. The output `next_tokens` gets incorrectly assigned to `past_key.0_RetainedState`. I'll avoid this change.

**Comment by quic-hemagnih** [comment]
Can you please rebase the code and see if there are any conflicts. Mostly code looks okay to me. We can plan its merge once all the pending comments are addressed

File: QEfficient/transformers/sampler/sampler.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+    repetition_penalties: Optional[torch.Tensor] = None,
+    past_presence_penalty_buffer: Optional[torch.Tensor] = None,
+    presence_penalties: Optional[torch.Tensor] = None,
+    temperatures: Optional[torch.Tensor] = None,
+    top_ks: Optional[torch.Tensor] = None,
+    top_ps: Optional[torch.Tensor] = None,
+    min_ps: Optional[torch.Tensor] = None,
+    random_numbers: Optional[torch.Tensor] = None,
+) -> Union[Tuple, SamplerOutput]:
+    r"""
```
Comment (quic-amitraj) [inline QEfficient/transformers/sampler/sampler.py:115]: Add docstring about the work of this new forward.

**Review by quic-amitraj — CHANGES_REQUESTED** [review]
Please add docstring to all the functions and classes.

File: QEfficient/transformers/sampler/sampler.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
+from dataclasses import dataclass
```
[the remark this answers is not in the document]
Comment (quic-sanising) [reply QEfficient/transformers/sampler/sampler.py:8]: Done

File: QEfficient/utils/constants.py [hunk 12]
```diff
@@ -56,10 +56,11 @@ def get_models_dir():
 
 QEFF_MODELS_DIR = get_models_dir()
 
-ONNX_EXPORT_EXAMPLE_BATCH_SIZE = 1
+ONNX_EXPORT_EXAMPLE_BATCH_SIZE = 2
```
[the remark this answers is not in the document]
Comment (quic-sanising) [reply QEfficient/utils/constants.py]: This is fixed

File: QEfficient/transformers/sampler/sampler.py (hunk 10 above)
Comment (quic-sanising) [reply QEfficient/transformers/sampler/sampler.py:122]: The doc string is available at `SamplerTransform` (https://github.com/quic-sanising/efficient-transformers/blob/on-device-sampling/QEfficient/transformers/models/pytorch_transforms.py#L460). This will be duplicated.

Comment (quic-sanising) [reply QEfficient/transformers/sampler/sampler.py:128]: Done

**Comment by quic-sanising** [comment]
> @quic-sanising Could you please add the description for this PR?

Done

**Comment by quic-hemagnih** [comment]
As discussed in the meeting, we will approve and merge this PR. Please have another PR with unit test added for this feature to ensure that its not broken by any subsequent checkins

**Review by quic-hemagnih — APPROVED** [review]

**Comment by quic-sanising** [comment]
Moved to #440
